src/lib.rs

```diff
@@ -481,26 +482,65 @@ const INTERNAL_ERROR_MSG: &str = "Fatal internal error. Please consider filing a
                                   report at https://github.com/clap-rs/clap/issues";
 const INVALID_UTF8: &str = "unexpected invalid UTF-8 code point";
 
-/// @TODO @release @docs
-pub trait Clap: From<ArgMatches> + IntoApp + Sized {}
+/// This trait is just a convenience on top of FromArgMatches + IntoApp
+pub trait Clap: FromArgMatches + IntoApp + Sized {
+    /// Parse from `std::env::args()`, exit on error
+    fn parse() -> Self {
+        let matches = <Self as IntoApp>::into_app().get_matches();
+        <Self as FromArgMatches>::from_arg_matches(&matches)
+    }
 
-/// @TODO @release @docs
-pub trait FromArgMatches: Sized {
+    /// Parse from `std::env::args()`, return Err on error.
+    fn try_parse() -> StdResult<Self, Error> {
+        let matches = <Self as IntoApp>::into_app().try_get_matches()?;
+        Ok(<Self as FromArgMatches>::from_arg_matches(&matches))
+    }
+
+    /// Parse from iterator, exit on error
+    fn parse_from<I, T>(itr: I) -> Self
+    where
+        I: IntoIterator<Item = T>,
+        // TODO (@CreepySkeleton): discover a way to avoid cloning here
+        T: Into<OsString> + Clone,
+    {
+        let matches = <Self as IntoApp>::into_app().get_matches_from(itr);
+        <Self as FromArgMatches>::from_arg_matches(&matches)
+    }
+
+    /// Parse from `std::env::args()`, return Err on error.
+    fn try_parse_from<I, T>(itr: I) -> StdResult<Self, Error>
+    where
+        I: IntoIterator<Item = T>,
+        // TODO (@CreepySkeleton): discover a way to avoid cloning here
+        T: Into<OsString> + Clone,
+    {
+        let matches = <Self as IntoApp>::into_app().try_get_matches_from(itr)?;
+        Ok(<Self as FromArgMatches>::from_arg_matches(&matches))
+    }
+}
+
+/// Build an App according to the struct
+///
+/// Also serves for flattening
+pub trait IntoApp: Sized {
     /// @TODO @release @docs
-    fn from_argmatches(matches: &crate::parse::ArgMatches) -> Self;
+    fn into_app<'b>() -> App<'b>;
+    /// @TODO @release @docs
+    fn augment_clap<'b>(app: App<'b>) -> App<'b>;
+}
 
+/// Extract values from ArgMatches into the struct.
+pub trait FromArgMatches: Sized {
     /// @TODO @release @docs
-    fn try_from_argmatches(
-        matches: &crate::parse::ArgMatches,
-    ) -> StdResult<Self, crate::parse::errors::Error> {
-        Ok(<Self as FromArgMatches>::from_argmatches(matches))
-    }
+    fn from_arg_matches<'b>(matches: &ArgMatches) -> Self;
 }
 
 /// @TODO @release @docs
-pub trait IntoApp: Sized {
+pub trait Subcommand: Sized {
+    /// @TODO @release @docs
+    fn from_subcommand<'b>(name: &str, matches: Option<&ArgMatches>) -> Option<Self>;
     /// @TODO @release @docs
-    fn into_app<'b>() -> crate::build::App<'b>;
+    fn augment_subcommands<'b>(app: App<'b>) -> App<'b>;
 }
```

Comment:
Can we move all these traits to `clap_derive`?

Reply:
No, `proc-macro = true` cannot export anything but proc macros. Why do you ask?